File: src/project_manager_flutter/lib/features/teams/repository/team_repository.dart
import 'package:dio/dio.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_exceptions.dart';
import '../models/add_member_request.dart';
import '../models/add_member_response.dart';
import '../models/create_team_request.dart';
import '../models/team.dart';
import '../models/team_detail.dart';

class TeamRepository {
  final Dio _dio;

  TeamRepository(this._dio);

  Future<List<Team>> listTeams(String projectId) async {
    try {
      final response = await _dio.get(
        ApiConstants.teams,
        queryParameters: {'projectId': projectId},
      );
      return (response.data as List).map((e) => Team.fromJson(e)).toList();
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<TeamDetail> getTeam(String id) async {
    try {
      final response = await _dio.get(ApiConstants.teamById(id));
      return TeamDetail.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> createTeam(CreateTeamRequest request) async {
    try {
      await _dio.post(ApiConstants.teams, data: request.toJson());
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<AddMemberResponse> addMember(String teamId, AddMemberRequest request) async {
    try {
      final response = await _dio.post(
        ApiConstants.teamMembers(teamId),
        data: request.toJson(),
      );
      return AddMemberResponse.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<void> removeMember(String teamId, String userId) async {
    try {
      await _dio.delete(ApiConstants.teamMember(teamId, userId));
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }
}

File: src/project-manager-web/src/features/teams/repository/team-repository.ts
import { apiClient } from "@/core/api/client";
import { ApiEndpoints } from "@/core/api/constants";
import { ApiException } from "@/core/api/exceptions";
import type { Team, TeamDetail, CreateTeamRequest, AddTeamMemberRequest } from "../models/types";
import { AxiosError } from "axios";

function handleError(error: unknown): never {
  if (error instanceof AxiosError) {
    throw ApiException.fromResponse(error.response?.status ?? 500, error.response?.data);
  }
  throw error;
}

export const teamRepository = {
  async list(projectId?: string): Promise<Team[]> {
    try {
      const response = await apiClient.get<Team[]>(ApiEndpoints.teams, {
        params: projectId ? { projectId } : undefined,
      });
      return response.data;
    } catch (error) { handleError(error); }
  },

  async get(id: string): Promise<TeamDetail> {
    try {
      const response = await apiClient.get<TeamDetail>(ApiEndpoints.teamById(id));
      return response.data;
    } catch (error) { handleError(error); }
  },

  async create(request: CreateTeamRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.teams, request);
    } catch (error) { handleError(error); }
  },

  async addMember(teamId: string, request: AddTeamMemberRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.teamMembers(teamId), request);
    } catch (error) { handleError(error); }
  },

  async removeMember(teamId: string, userId: string): Promise<void> {
    try {
      await apiClient.delete(ApiEndpoints.teamMember(teamId, userId));
    } catch (error) { handleError(error); }
  },
};
